language/move-vm/runtime/src/interpreter.rs

```diff
@@ -262,15 +262,38 @@ impl Interpreter {
     ///
     /// Native functions do not push a frame at the moment and as such errors from a native
     /// function are incorrectly attributed to the caller.
-    fn make_call_frame(&mut self, func: Arc<Function>, ty_args: Vec<Type>) -> VMResult<Frame> {
+    fn make_call_frame<'a>(
+        &mut self,
+        func: Arc<Function>,
+        ty_args: Vec<Type>,
+        loader: &'a Loader,
+    ) -> VMResult<Frame> {
+        let resolver = func.get_resolver(loader);
         let mut locals = Locals::new(func.local_count());
         let arg_count = func.arg_count();
         for i in 0..arg_count {
+            let v = self.operand_stack.pop().map_err(|e| self.set_location(e))?;
+            let expected_ty = resolver
+                .resolve_signature_token(&func.local_types()[arg_count - i - 1])
+                .and_then(|ty| ty.subst(&ty_args))
+                .map_err(|e| match func.module_id() {
+                    Some(id) => e
+                        .at_code_offset(func.index(), 0)
+                        .finish(Location::Module(id.clone())),
+                    None => {
+                        let err =
+                            PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                                .with_message(
+                                    "Unexpected native function not located in a module".to_owned(),
+                                );
+                        self.set_location(err)
+                    }
+                })?;
+
+            v.check_type(&expected_ty)
```

Comment:
As I commented under `fn check_type`, this operation performs a deep traversal over the whole value and is very concerning in terms of performance and gas

Reply:
Yeah I guess I'm a little bit torn on this... On one hand, I understand that if you want to be absolutely secure, you need to check the whole value recursively across function boundaries. On the other hand, I'm also really concerned about the perf penalty we're incurring here.

cc @wrwg